// @strapi/admin/dist/admin/admin/src/pages/Home/components/ContentManagerWidgets.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import { Typography, Table, Tbody, Tr, Td, Box, IconButton, Status } from '@strapi/design-system';
import { Pencil, CheckCircle } from '@strapi/icons';
import { useIntl } from 'react-intl';
import { useNavigate, Link } from 'react-router-dom';
import { styled } from 'styled-components';
import { RelativeTime } from '../../../components/RelativeTime.mjs';
import { useTracking } from '../../../features/Tracking.mjs';
import { useGetRecentDocumentsQuery } from '../../../services/homepage.mjs';
import { capitalise } from '../../../utils/strings.mjs';
import { Widget } from './Widget.mjs';

const CellTypography = styled(Typography).attrs({
    maxWidth: '14.4rem',
    display: 'block'
})`
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;
const DocumentStatus = ({ status = 'draft' })=>{
    const statusVariant = status === 'draft' ? 'secondary' : status === 'published' ? 'success' : 'alternative';
    const { formatMessage } = useIntl();
    return /*#__PURE__*/ jsx(Status, {
        variant: statusVariant,
        size: "XS",
        children: /*#__PURE__*/ jsx(Typography, {
            tag: "span",
            variant: "omega",
            fontWeight: "bold",
            children: formatMessage({
                id: `content-manager.containers.List.${status}`,
                defaultMessage: capitalise(status)
            })
        })
    });
};
const RecentDocumentsTable = ({ documents })=>{
    const { formatMessage } = useIntl();
    const { trackUsage } = useTracking();
    const navigate = useNavigate();
    const getEditViewLink = (document)=>{
        // TODO: import the constants for this once the code is moved to the CM package
        const isSingleType = document.kind === 'singleType';
        const kindPath = isSingleType ? 'single-types' : 'collection-types';
        const queryParams = document.locale ? `?plugins[i18n][locale]=${document.locale}` : '';
        return `/content-manager/${kindPath}/${document.contentTypeUid}${isSingleType ? '' : '/' + document.documentId}${queryParams}`;
    };
    const handleRowClick = (document)=>()=>{
            trackUsage('willEditEntryFromHome');
            const link = getEditViewLink(document);
            navigate(link);
        };
    return /*#__PURE__*/ jsx(Table, {
        colCount: 5,
        rowCount: documents?.length ?? 0,
        children: /*#__PURE__*/ jsx(Tbody, {
            children: documents?.map((document)=>/*#__PURE__*/ jsxs(Tr, {
                    onClick: handleRowClick(document),
                    cursor: "pointer",
                    children: [
                        /*#__PURE__*/ jsx(Td, {
                            children: /*#__PURE__*/ jsx(CellTypography, {
                                title: document.title,
                                variant: "omega",
                                textColor: "neutral800",
                                children: document.title
                            })
                        }),
                        /*#__PURE__*/ jsx(Td, {
                            children: /*#__PURE__*/ jsx(CellTypography, {
                                variant: "omega",
                                textColor: "neutral600",
                                children: document.kind === 'singleType' ? formatMessage({
                                    id: 'content-manager.widget.last-edited.single-type',
                                    defaultMessage: 'Single-Type'
                                }) : formatMessage({
                                    id: document.contentTypeDisplayName,
                                    defaultMessage: document.contentTypeDisplayName
                                })
                            })
                        }),
                        /*#__PURE__*/ jsx(Td, {
                            children: /*#__PURE__*/ jsx(Box, {
                                display: "inline-block",
                                children: document.status ? /*#__PURE__*/ jsx(DocumentStatus, {
                                    status: document.status
                                }) : /*#__PURE__*/ jsx(Typography, {
                                    textColor: "neutral600",
                                    "aria-hidden": true,
                                    children: "-"
                                })
                            })
                        }),
                        /*#__PURE__*/ jsx(Td, {
                            children: /*#__PURE__*/ jsx(Typography, {
                                textColor: "neutral600",
                                children: /*#__PURE__*/ jsx(RelativeTime, {
                                    timestamp: new Date(document.updatedAt)
                                })
                            })
                        }),
                        /*#__PURE__*/ jsx(Td, {
                            onClick: (e)=>e.stopPropagation(),
                            children: /*#__PURE__*/ jsx(Box, {
                                display: "inline-block",
                                children: /*#__PURE__*/ jsx(IconButton, {
                                    tag: Link,
                                    to: getEditViewLink(document),
                                    onClick: ()=>trackUsage('willEditEntryFromHome'),
                                    label: formatMessage({
                                        id: 'content-manager.actions.edit.label',
                                        defaultMessage: 'Edit'
                                    }),
                                    variant: "ghost",
                                    children: /*#__PURE__*/ jsx(Pencil, {})
                                })
                            })
                        })
                    ]
                }, document.documentId))
        })
    });
};
/* -------------------------------------------------------------------------------------------------
 * LastEditedWidget
 * -----------------------------------------------------------------------------------------------*/ const LastEditedWidgetContent = ()=>{
    const { formatMessage } = useIntl();
    const { data, isLoading, error } = useGetRecentDocumentsQuery({
        action: 'update'
    });
    if (isLoading) {
        return /*#__PURE__*/ jsx(Widget.Loading, {});
    }
    if (error || !data) {
        return /*#__PURE__*/ jsx(Widget.Error, {});
    }
    if (data.length === 0) {
        return /*#__PURE__*/ jsx(Widget.NoData, {
            children: formatMessage({
                id: 'content-manager.widget.last-edited.no-data',
                defaultMessage: 'No edited entries'
            })
        });
    }
    return /*#__PURE__*/ jsx(RecentDocumentsTable, {
        documents: data
    });
};
const LastEditedWidget = ()=>{
    return /*#__PURE__*/ jsx(Widget.Root, {
        title: {
            id: 'content-manager.widget.last-edited.title',
            defaultMessage: 'Last edited entries'
        },
        icon: Pencil,
        children: /*#__PURE__*/ jsx(LastEditedWidgetContent, {})
    });
};
/* -------------------------------------------------------------------------------------------------
 * LastPublishedWidget
 * -----------------------------------------------------------------------------------------------*/ const LastPublishedWidgetContent = ()=>{
    const { formatMessage } = useIntl();
    const { data, isLoading, error } = useGetRecentDocumentsQuery({
        action: 'publish'
    });
    if (isLoading) {
        return /*#__PURE__*/ jsx(Widget.Loading, {});
    }
    if (error || !data) {
        return /*#__PURE__*/ jsx(Widget.Error, {});
    }
    if (data.length === 0) {
        return /*#__PURE__*/ jsx(Widget.NoData, {
            children: formatMessage({
                id: 'content-manager.widget.last-published.no-data',
                defaultMessage: 'No published entries'
            })
        });
    }
    return /*#__PURE__*/ jsx(RecentDocumentsTable, {
        documents: data
    });
};
const LastPublishedWidget = ()=>{
    return /*#__PURE__*/ jsx(Widget.Root, {
        title: {
            id: 'content-manager.widget.last-published.title',
            defaultMessage: 'Last published entries'
        },
        icon: CheckCircle,
        children: /*#__PURE__*/ jsx(LastPublishedWidgetContent, {})
    });
};

export { LastEditedWidget, LastPublishedWidget };
