// @strapi/content-manager/dist/admin/pages/EditView/components/FormInputs/Wysiwyg/EditorLayout.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import * as React from 'react';
import { Flex, Box, Button, Portal, FocusTrap, Typography } from '@strapi/design-system';
import { Collapse } from '@strapi/icons';
import { useIntl } from 'react-intl';
import { styled } from 'styled-components';
import { PreviewWysiwyg } from './PreviewWysiwyg.mjs';

const EditorLayout = ({ children, isExpandMode, error, previewContent = '', onCollapse })=>{
    const { formatMessage } = useIntl();
    React.useEffect(()=>{
        if (isExpandMode) {
            document.body.classList.add('lock-body-scroll');
        }
        return ()=>{
            document.body.classList.remove('lock-body-scroll');
        };
    }, [
        isExpandMode
    ]);
    if (isExpandMode) {
        return /*#__PURE__*/ jsx(Portal, {
            role: "dialog",
            "aria-modal": false,
            children: /*#__PURE__*/ jsx(FocusTrap, {
                onEscape: onCollapse,
                children: /*#__PURE__*/ jsx(ExpandWrapper, {
                    position: "fixed",
                    top: 0,
                    left: 0,
                    right: 0,
                    bottom: 0,
                    zIndex: 4,
                    justifyContent: "center",
                    onClick: onCollapse,
                    children: /*#__PURE__*/ jsx(Box, {
                        background: "neutral0",
                        hasRadius: true,
                        shadow: "popupShadow",
                        overflow: "hidden",
                        width: "90%",
                        height: "90%",
                        onClick: (e)=>e.stopPropagation(),
                        children: /*#__PURE__*/ jsxs(Flex, {
                            height: "100%",
                            alignItems: "flex-start",
                            children: [
                                /*#__PURE__*/ jsx(BoxWithBorder, {
                                    flex: "1",
                                    height: "100%",
                                    children: children
                                }),
                                /*#__PURE__*/ jsxs(Flex, {
                                    alignItems: "start",
                                    direction: "column",
                                    flex: 1,
                                    height: "100%",
                                    width: "100%",
                                    children: [
                                        /*#__PURE__*/ jsx(Flex, {
                                            height: "4.8rem",
                                            background: "neutral100",
                                            justifyContent: "flex-end",
                                            shrink: 0,
                                            width: "100%",
                                            children: /*#__PURE__*/ jsxs(ExpandButton, {
                                                onClick: onCollapse,
                                                variant: "tertiary",
                                                size: "M",
                                                children: [
                                                    /*#__PURE__*/ jsx(Typography, {
                                                        children: formatMessage({
                                                            id: 'components.Wysiwyg.collapse',
                                                            defaultMessage: 'Collapse'
                                                        })
                                                    }),
                                                    /*#__PURE__*/ jsx(Collapse, {})
                                                ]
                                            })
                                        }),
                                        /*#__PURE__*/ jsx(Box, {
                                            position: "relative",
                                            height: "100%",
                                            width: "100%",
                                            children: /*#__PURE__*/ jsx(PreviewWysiwyg, {
                                                data: previewContent
                                            })
                                        })
                                    ]
                                })
                            ]
                        })
                    })
                })
            })
        });
    }
    return /*#__PURE__*/ jsx(Flex, {
        borderColor: error ? 'danger600' : 'neutral200',
        borderStyle: "solid",
        borderWidth: "1px",
        hasRadius: true,
        direction: "column",
        alignItems: "stretch",
        children: children
    });
};
const ExpandWrapper = styled(Flex)`
  background: ${({ theme })=>`${theme.colors.neutral800}${Math.floor(0.2 * 255).toString(16).padStart(2, '0')}`};
`;
const BoxWithBorder = styled(Box)`
  border-right: 1px solid ${({ theme })=>theme.colors.neutral200};
`;
const ExpandButton = styled(Button)`
  background-color: transparent;
  border: none;
  align-items: center;

  & > span {
    display: flex;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    font-weight: ${({ theme })=>theme.fontWeights.regular};
  }

  svg {
    margin-left: ${({ theme })=>`${theme.spaces[2]}`};

    path {
      fill: ${({ theme })=>theme.colors.neutral700};
      width: 1.2rem;
      height: 1.2rem;
    }
  }
`;

export { EditorLayout, ExpandButton };
